src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs

```diff
@@ -1991,29 +1996,301 @@ static void Main()
+                //         ref var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "delegate () { }").WithLocation(7, 22));
+        }
+
+        [Fact]
+        public void ImplicitlyTypedVariables_04()
+        {
+            var source =
+@"class Program
+{
+    static void Main()
+    {
+        using var d1 = Main;
+        using var d2 = () => { };
+        using var d3 = delegate () { };
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (5,9): error CS1674: 'Action': type used in a using statement must be implicitly convertible to 'System.IDisposable'.
+                //         using var d1 = Main;
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "using var d1 = Main;").WithArguments("System.Action").WithLocation(5, 9),
+                // (6,9): error CS1674: 'Action': type used in a using statement must be implicitly convertible to 'System.IDisposable'.
+                //         using var d2 = () => { };
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "using var d2 = () => { };").WithArguments("System.Action").WithLocation(6, 9),
+                // (7,9): error CS1674: 'Action': type used in a using statement must be implicitly convertible to 'System.IDisposable'.
+                //         using var d3 = delegate () { };
+                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "using var d3 = delegate () { };").WithArguments("System.Action").WithLocation(7, 9));
+        }
+
+        [Fact]
+        public void ImplicitlyTypedVariables_05()
+        {
+            var source =
+@"class Program
+{
+    static void Main()
+    {
+        foreach (var d1 in Main) { }
+        foreach (var d2 in () => { }) { }
+        foreach (var d3 in delegate () { }) { }
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (5,28): error CS0446: Foreach cannot operate on a 'method group'. Did you intend to invoke the 'method group'?
+                //         foreach (var d1 in Main) { }
+                Diagnostic(ErrorCode.ERR_AnonMethGrpInForEach, "Main").WithArguments("method group").WithLocation(5, 28),
+                // (6,28): error CS0446: Foreach cannot operate on a 'lambda expression'. Did you intend to invoke the 'lambda expression'?
+                //         foreach (var d2 in () => { }) { }
+                Diagnostic(ErrorCode.ERR_AnonMethGrpInForEach, "() => { }").WithArguments("lambda expression").WithLocation(6, 28),
+                // (7,28): error CS0446: Foreach cannot operate on a 'anonymous method'. Did you intend to invoke the 'anonymous method'?
+                //         foreach (var d3 in delegate () { }) { }
+                Diagnostic(ErrorCode.ERR_AnonMethGrpInForEach, "delegate () { }").WithArguments("anonymous method").WithLocation(7, 28));
+        }
```

Comment:
nit: consider another similar negative test with deconstruction declaration `(var x, var y) = ...`

Reply:
Also, something like `(Action a, var b) = (null, () => {});` (I assume should work per spec, but may not work in current implementation)

Reply:
Please check `GetTypeInfo` and `GetDeclaredSymbol` in one of the positive tests

Reply:
Consider including a test with a non-simple signature (error)

Reply:
Added tests. I'll investigate the failures with the tuple cases.